contracts/PionexContract.sol

```diff
@@ -0,0 +1,366 @@
+        require(_fill.expiry > uint64(block.timestamp), "PionexContract: Fill request is expired");
+        require(_fill.recipient != address(0), "PionexContract: recipient can not be zero address");
+
+        bytes32 fillHash = getEIP712Hash(PionexContractLibEIP712._getFillStructHash(_fill));
+        require(isValidSignature(_fill.pionex, fillHash, bytes(""), _fillTakerSig), "PionexContract: Fill is not signed by pionex");
+
+        // Set fill seen to avoid replay attack.
+        // PermanentStorage would throw error if fill is already seen.
+        permStorage.setLimitOrderTransactionSeen(fillHash);
+    }
+
+    function _validateFillPermission(
+        bytes32 _orderHash,
+        uint256 _fillAmount,
+        address _executor,
+        CoordinatorParams memory _crdParams
+    ) internal returns (bytes32) {
+        require(_crdParams.expiry > uint64(block.timestamp), "PionexContract: Fill permission is expired");
+
+        bytes32 allowFillHash = getEIP712Hash(
+            PionexContractLibEIP712._getAllowFillStructHash(
+                PionexContractLibEIP712.AllowFill({
+                    orderHash: _orderHash,
+                    executor: _executor,
+                    fillAmount: _fillAmount,
+                    salt: _crdParams.salt,
+                    expiry: _crdParams.expiry
+                })
+            )
+        );
+        require(isValidSignature(coordinator, allowFillHash, bytes(""), _crdParams.sig), "PionexContract: AllowFill is not signed by coordinator");
+
+        // Set allow fill seen to avoid replay attack
+        // PermanentStorage would throw error if allow fill is already seen.
+        permStorage.setLimitOrderAllowFillSeen(allowFillHash);
+
+        return allowFillHash;
+    }
+
+    struct TraderSettlement {
+        bytes32 orderHash;
+        bytes32 allowFillHash;
+        address trader;
+        address recipient;
+        address user;
+        address pionex;
+        IERC20 userToken;
+        IERC20 pionexToken;
+        uint256 userTokenAmount;
+        uint256 pionexTokenAmount;
+        uint256 remainingUserTokenAmount;
+        uint16 gasFeeFactor;
+        uint16 pionexStrategyFeeFactor;
+    }
+
+    function _settleForTrader(TraderSettlement memory _settlement) internal returns (uint256) {
+        // memory cache
+        ISpender _spender = spender;
+        address _feeCollector = feeCollector;
+
+        // Calculate user fee (user receives pionex token so fee is charged in pionex token)
+        // 1. Fee for Tokenlon
+        uint256 tokenlonFee = _mulFactor(_settlement.pionexTokenAmount, tokenlonFeeFactor);
+        // 2. Fee for Pionex, including gas fee and strategy fee
+        uint256 pionexFee = _mulFactor(_settlement.pionexTokenAmount, _settlement.gasFeeFactor + _settlement.pionexStrategyFeeFactor);
+        uint256 pionexTokenForMaker = _settlement.pionexTokenAmount.sub(tokenlonFee).sub(pionexFee);
```

Comment:
Q: Would it possible that token after deducting fees makes the final exchange ratio less than `_order.pionexTokenAmount/_order.userTokenAmount`?

Reply:
Yeah it is possible. We haven't decide on whether we should enforce that the eventual ratio is at least as good as the one user specified in the order. Since the gas fee will inevitably reduce the ratio and if Pionex provide the same ratio as the one specified by user, this would result in the eventual ratio worse than the user specified.